theano/sandbox/mkl/mkl_type.py

```diff
@@ -0,0 +1,270 @@
+from __future__ import absolute_import, print_function, division
+import os
+import numpy
+import theano
+from theano import Type, Variable
+from theano import tensor
+from six import StringIO
+
+try:
+    import mkl_ndarray
+    import mkl_ndarray.mkl_ndarray as mkl
+
+    try:
+        mkl_ndarray.__file__
+    except AttributeError:
+        from theano.gof.cmodule import get_lib_extension
+        mkl_ndarray.__file__ = os.path.join(mkl_ndarray.__path__._path[0], 'mkl_ndarray.' + get_lib_extension())
+except ImportError:
+    mkl = None
+
+
+class _operators(tensor.basic._tensor_py_operators):
+    dtype = property(lambda s: s.type.dtype)
+    broadcastable = property(lambda s: s.type.broadcastable)
+    ndim = property(lambda s: s.type.ndim)
+
+
+class MKLNdarrayVariable(_operators, Variable):
+    pass
+
+
+class MKLNdarrayType(Type):
```

Comment:
I'm not sure I get everything. By layout, what do you mean? I see 2 interpretations that could be used here that are close to each others and used in Theano.
1) The memory layout. specificed at 100% by shapes and strides. It don't give any semantic to each dimensions, just how to read the data in the right order.
2) The "semantic layout". Is the layout for the inputs images (batch, in channel, image height, image width) or the other semantic layout as tensorflow, cuda convnet and nervana use different layout.

In one, do you mean that your C object have the "semantic layout" inside it?

I checked rapidly the URL you gave. A small comments. I saw many fct that ends with F32 or F64. I suppose it mean float32 or float64. Can you confirm?